pyccel/codegen/printing/fcode.py

```diff
@@ -612,25 +610,161 @@ def _print_PythonImag(self, expr):
-        return expr.fprint(self._print)
+        """Fortran print."""
+        a_code = self._print(expr.a)
+        b_code = self._print(expr.b)
+
+        if expr.a.order and expr.b.order:
+            if expr.a.order != expr.b.order:
+                raise NotImplementedError("Mixed order matmul not supported.")
+
+        # Fortran ordering
+        if expr.a.order == 'F':
+            return 'matmul({0},{1})'.format(a_code, b_code)
+
+        # C ordering
+        return 'matmul({1},{0})'.format(a_code, b_code)
 
     def _print_NumpyCross(self, expr):
-        return expr.fprint(self._print)
+        """Fortran print."""
+
+        a     = IndexedVariable(expr.first)
+        b     = IndexedVariable(expr.second)
+        slc   = Slice(None, None)
+        rank  = expr.rank
+
+        if rank > 2:
+            raise NotImplementedError('TODO')
+
+        if rank == 2:
+            a_inds = [[slc,0], [slc,1], [slc,2]]
+            b_inds = [[slc,0], [slc,1], [slc,2]]
+
+            if expr.first.order == 'C':
+                for inds in a_inds:
+                    inds.reverse()
+            if expr.second.order == 'C':
+                for inds in b_inds:
+                    inds.reverse()
+
+            a = [a[tuple(inds)] for inds in a_inds]
+            b = [b[tuple(inds)] for inds in b_inds]
+
+
+        cross_product = [a[1]*b[2]-a[2]*b[1],
+                         a[2]*b[0]-a[0]*b[2],
+                         a[0]*b[1]-a[1]*b[0]]
+
+        cross_product = PythonTuple(cross_product)
+        cross_product = self._print(cross_product)
+        first = self._print(expr.first)
+        order = expr.order
+
+
+        if rank == 2:
+
+            if order == 'C':
+
+                code = 'reshape({}, shape({}), order=[2, 1])'.format(cross_product, first)
+            else:
+
+                code = 'reshape({}, shape({})'.format(cross_product, first)
+
+        elif rank == 1:
+            code = cross_product
+
+        return code
+
+    def _print_NumpyDiag(self, expr, lhs):
+        """Fortran print."""
+
+        array = self._print(expr.array)
+        rank  = expr.array.rank
+
+        if rank == 2:
+            lhs   = IndexedVariable(lhs)[expr.index]
+            rhs   = IndexedVariable(expr.array)[expr.index,expr.index]
+            body  = [Assign(lhs, rhs)]
+            body  = For(expr.index, PythonRange(PythonLen(expr.array)), body)
+            code  = self._print(body)
+            alloc = 'allocate({0}(0: size({1},1)-1))'.format(lhs.base, array)
+        elif rank == 1:
+
+            lhs   = IndexedVariable(lhs)[expr.index, expr.index]
+            rhs   = IndexedVariable(expr.array)[expr.index]
+            body  = [Assign(lhs, rhs)]
+            body  = For(expr.index, PythonRange(PythonLen(expr.array)), body)
+            code  = self._print(body)
+            alloc = 'allocate({0}(0: size({1},1)-1, 0: size({1},1)-1))'.format(lhs, array)
+
+        return alloc + '\n' + code
```

Comment:
Didn't you delete this class?

Reply:
this is not a class, but yes I removed `NumpyDiag` in ast.numpyast, and sorry for `lhs` argument I think its cursed for me.

Reply:
You are right, it is not a class, but it is a function that is only called on instances of the class `NumpyDiag` so if `NumpyDiag` doesn't exist anymore then this function doesn't need to either

Reply:
The `lhs` being here is the reason why we said to delete this class. There is no way round it. It's not your fault ^^